middleware/plugin/middleware.go

```diff
@@ -0,0 +1,140 @@
+//go:build plugin
+
+package plugin
+
+/*
+ * SPDX-FileCopyrightText: 2023 Siemens AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Author: Michael Adler <michael.adler@siemens.com>
+ */
+
+import (
+	"net/http"
+	"net/url"
+
+	"github.com/Southclaws/fault"
+	"github.com/rs/zerolog/log"
+	genPlugin "github.com/siemens/wfx/generated/plugin"
+	"github.com/siemens/wfx/generated/plugin/client"
+	"github.com/siemens/wfx/middleware/logging"
+)
+
+type MW struct {
+	plugin Plugin
+}
+
+func NewMiddleware(plugin Plugin) (*MW, error) {
+	log.Debug().Str("plugin", plugin.Name()).Msg("Creating new plugin middleware")
+
+	if err := plugin.Start(); err != nil {
+		_ = plugin.Wait()
+		log.Err(err).Msg("Failed to start plugin")
+		return nil, fault.Wrap(err)
+	}
+	mw := MW{plugin: plugin}
+
+	go func() {
+		for {
+			p := mw.plugin
+
+			err := p.Wait()
+			log.Warn().Err(err).Msg("Plugin stopped")
+			if p.IsStopped() {
+				log.Info().Msg("Plugin was stopped, not restarting it")
+				break
+			}
+
+			if err := p.Start(); err != nil {
+				log.Err(err).Msg("Failed to start plugin")
```

Comment:
If a plugin doesn't start that should be a fatal error, no? After all, it's there for a reason?

Reply:
Well actually it should be attempted to restart, since this goroutine is the supervisor. I have adjusted the code slightly to deal with this scenario.